ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/result/impl/PreventiveAndCurativesRaoResultImpl.java

```diff
@@ -112,25 +115,81 @@ private static Map<State, PerimeterResult> buildPostContingencyResults(StateTree
+                results.put(contingencyScenario.getCurativeState(), new PerimeterResultImpl(preContingencyResult, curativeResult));
             }
         });
         return results;
     }
 
+    private Set<String> getContingenciesToExclude(StateTree stateTree) {
+        Set<String> contingenciesToExclude = new HashSet<>();
+        stateTree.getContingencyScenarios().forEach(contingencyScenario -> {
+            Optional<State> automatonState = contingencyScenario.getAutomatonState();
+            if (automatonState.isPresent()) {
+                OptimizationResult automatonResult = postContingencyResults.get(automatonState.get());
+                if (!automatonResult.getContingencies().contains(contingencyScenario.getContingency().getId())) {
+                    contingenciesToExclude.add(contingencyScenario.getContingency().getId());
+                    return;
+                }
+                OptimizationResult curativeResult = postContingencyResults.get(contingencyScenario.getCurativeState());
+                if (!curativeResult.getContingencies().contains(contingencyScenario.getContingency().getId())) {
+                    contingenciesToExclude.add(contingencyScenario.getContingency().getId());
+                }
+            } else {
+                OptimizationResult curativeResult = postContingencyResults.get(contingencyScenario.getCurativeState());
+                if (!curativeResult.getContingencies().contains(contingencyScenario.getContingency().getId())) {
+                    contingenciesToExclude.add(contingencyScenario.getContingency().getId());
+                }
+            }
+        });
+        return contingenciesToExclude;
+    }
+
+    private void excludeContingencies(Set<String> contingenciesToExclude) {
+        initialResult.excludeContingencies(contingenciesToExclude);
+        firstPreventivePerimeterResult.excludeContingencies(contingenciesToExclude);
+        secondPreventivePerimeterResult.excludeContingencies(contingenciesToExclude);
+        resultsWithPrasForAllCnecs.excludeContingencies(contingenciesToExclude);
+        postContingencyResults.values().forEach(result -> result.excludeContingencies(contingenciesToExclude));
+    }
+
     @Override
     public ComputationStatus getComputationStatus() {
         if (initialResult.getSensitivityStatus() == FAILURE
             || secondPreventivePerimeterResult.getSensitivityStatus() == FAILURE
-            || postContingencyResults.values().stream().anyMatch(perimeterResult -> perimeterResult.getSensitivityStatus() == FAILURE)) {
+            || postContingencyResults.values().stream().anyMatch(perimeterResult -> Objects.isNull(perimeterResult) || perimeterResult.getSensitivityStatus() == FAILURE)) {
             return FAILURE;
         }
         // TODO: specify the behavior in case some perimeter are FALLBACK and other ones DEFAULT
         return ComputationStatus.DEFAULT;
     }
 
+    @Override
```

Comment:
why not use afterOptimizing :         return getPerimeterResult(OptimizationState.afterOptimizing(state.getInstant()), state).getSensitivityStatus(state);